### ISSUE TITLE
NetMQ.FaultException: SocketBase.Recv passed an uninitialised Msg in NetMqPoller

### ISSUE THREAD

**Issue by ylexus** [zeromq/netmq#635]
### Environment

    NetMQ Version:    3.3.3.4
    Operating System: Windows 7
    .NET Version:     4.5.1

### Expected behaviour
This does not happen

### Actual behaviour
This and some other exception is thrown by NetMQ

### Steps to reproduce the behaviour
I use the abstraction below to implement a non-blocking async RPC connection. it's operated on two threads: "rpc-poller" runs the Run() method and "rpc-state" executes all other methods. I've seen the following occurring when I am frequently disposing instances of NetMqRpcConnection:

```
LOG4NET 2017-01-11 12:46:55,989 [rpc-poller] ERROR 
NetMQ.FaultException: SocketBase.Recv passed an uninitialised Msg.
   at NetMQ.Core.SocketBase.TryRecv(Msg& msg, TimeSpan timeout)
   at NetMQ.NetMQSocket.TryReceive(Msg& msg, TimeSpan timeout)
   at NetMQ.NetMQPoller.<.ctor>b__20_0(Object <sender>, NetMQQueueEventArgs`1 <e>)
   at NetMQ.EventDelegator`1.Fire(Object sender, T args)
   at NetMQ.NetMQQueue`1.OnReceiveReady(Object sender, NetMQSocketEventArgs e)
   at NetMQ.NetMQSocket.InvokeEvents(Object sender, PollEvents events)
   at NetMQ.NetMQPoller.Run()
   at ????.Rpc.NetMqRpcConnection.Run()
...
```

And 25ms later I get this entry:

```
LOG4NET 2017-01-11 12:46:56,014 [rpc-state] ERROR 
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator.MoveNext()
   at NetMQ.NetMQPoller.Dispose()
   at ????.Rpc.NetMqRpcConnection.Disconnect()
...
```

I suppose these are related. Why would this happen? Since I am not breaking any NetMQ threading rules, this looks like an internal NetMQ issue?

```java
    sealed class NetMqRpcConnection : IRpcConnection
    {
        static readonly ILog log = LogManager.GetLogger(typeof(NetMqRpcConnection));

        readonly string connectionString;
        readonly Action<RaptorRpcMessage> responseHandler;
        readonly TimeSpan requestTimeout;

        // state
        DealerSocket socket;
        NetMQQueue<ConnectionRequestContext> sendQueue;
        NetMQPoller poller;

        internal NetMqRpcConnection(string connectionString, TimeSpan requestTimeout, Action<RaptorRpcMessage> responseHandler)
        {
            this.connectionString = CheckNotNull(connectionString);
            this.requestTimeout = CheckNotNull(requestTimeout);
            this.responseHandler = CheckNotNull(responseHandler);
        }

        public void Connect()
        {
            // we should ensure to send and receive from the socket on the same thread - this is the NetMQ requirement
            // typical way of doing this is to use NetMQQueue to queue requests and use the same poller to poll this queue and the socket
            // see https://github.com/zeromq/netmq/issues/441
            socket = new DealerSocket();
            socket.Options.Linger = TimeSpan.FromSeconds(0);
            socket.Options.Identity = Encoding.Unicode.GetBytes(RpcGuid.New());
            socket.ReceiveReady += (sender, args) => onReceive(args);
            socket.Connect(connectionString);
            sendQueue = new NetMQQueue<ConnectionRequestContext>();
            sendQueue.ReceiveReady += (sender, args) => onSend(args);
            poller = new NetMQPoller { socket, sendQueue };
        }

        public void Run()
        {
            poller.Run();
        }

        public void Send(RaptorRpcMessage request, Action sendSuccessHandler, Action sendTimeoutHandler)
        {
            sendQueue.Enqueue(new ConnectionRequestContext(request, sendSuccessHandler, sendTimeoutHandler));
        }

        public void Disconnect()
        {
            if (poller != null)
            {
                poller.Dispose();
                poller = null;

            }
            if (socket != null)
            {
                socket.Dispose();
                socket = null;
            }
        }

        void onSend(NetMQQueueEventArgs<ConnectionRequestContext> args)
        {
            ConnectionRequestContext requestContext = args.Queue.Dequeue();
            NetMQMessage requestMsg = createNetMqMsg(requestContext.Request);
            log.InfoFormat("RPC {0}: OUT {1}", connectionString, requestMsg);

            if (socket.TrySendMultipartMessage(requestTimeout, requestMsg))
            {
                requestContext.SendSuccessHandler.Invoke();
            }
            else
            {
                requestContext.SendTimeoutHandler.Invoke();
            }
        }

        void onReceive(NetMQSocketEventArgs args)
        {
            NetMQMessage response = new NetMQMessage();
            args.Socket.TryReceiveMultipartMessage(ref response);
            log.InfoFormat("RPC {0}:  IN {1}", connectionString, response);

            if (response.FrameCount != 2)
            {
                log.ErrorFormat("RPC {0}: Wrong frame count ({1}) on incoming message {2}", connectionString, response.FrameCount, response);
            }
            else
            {
                string requestId = response[0].ConvertToString();
                string responsePayload = response[1].ConvertToString();
                responseHandler.Invoke(new RaptorRpcMessage(requestId, responsePayload));
            }
        }

        static NetMQMessage createNetMqMsg(RaptorRpcMessage requestMessage)
        {
            var msg = new NetMQMessage();
            msg.Append(requestMessage.RequestId);
            msg.Append(requestMessage.Payload);
            return msg;
        }
    }
```

**Comment by goforgold** [comment]
@ylexus 

Did you find any solution to the issue? I am facing the same issue.

**Comment by somdoron** [comment]
@goforgold can you share more details? Stacktrace? Which API methods are you using for sending?

**Comment by somdoron** [comment]
and which version of netmq

**Comment by somdoron** [comment]
any chance you are trying to dequeue of after the queue is disposed?

**Comment by ylexus** [comment]
> @ylexus
> 
> Did you find any solution to the issue? I am facing the same issue.

Unfortunately I no longer work on that project or company, and don’t remember what workaround we found for this particular issue.

**Comment by goforgold** [comment]
> @goforgold can you share more details? Stacktrace? Which API methods are you using for sending?

Hi @somdoron 

Thanks for your response. Otherwise, by looking at the number of open issues without any single comment made me a bit sceptical whether to use this project or not. 

Actually, my issue got resolved. I was trying to follow a C# example from [Router-Dealer NetMQ documentation](https://netmq.readthedocs.io/en/latest/router-dealer/).

Below piece of code from example was causing a compile-time issue in 4.x version of the library.
```
e.Socket.Receive(out hasmore);
```

I changed it to
```
NetMQMessage msg = null;
e.Socket.Receive(msg);
```
Since the return type of `Receive` method here is `void` and parameter type is `ref` so I guessed that I will receive the message through the parameter and sent uninitialized instance of `NetMQMessage`.

But, I later changed it to below which most people seems to be doing nowadays 😉
```
NetMQMessage msg = e.Socket.ReceiveMultipartMessage();
```
It worked just as expected among many other things I tested before I decided to use the library.

Once again, really thanks for your response @somdoron.
